### PR TITLE
Upgrade to DF 16+ version

### PR DESCRIPTION
**Pull request by gruuya**
Upgrade Seafowl to Datafusion post 16.0.0 that includes the fix needed to convert statements to plan from the session state.

### CODE REVIEW

File: src/context.rs
Comment (neumark) [inline src/context.rs:573]: Due to a datafusion bug (still unfixed in 16.0), calling `ObjectName.to_string()` will double quote names with special chars. Instead, we must do:
```rust
pub fn remove_quotes_from_ident(possibly_quoted_name: &Ident) -> Ident {
    Ident::new(&possibly_quoted_name.value)
}

pub fn remove_quotes_from_idents(column_names: &[Ident]) -> Vec<Ident> {
    column_names.iter().map(remove_quotes_from_ident).collect()
}

pub fn remove_quotes_from_object_name(name: &ObjectName) -> ObjectName {
    ObjectName(remove_quotes_from_idents(&name.0))
}

// ...
let full_name = remove_quotes_from_object_name(&name).to_string()
```

These fuctions are added to `context.rs` in [PR 269](https://github.com/splitgraph/seafowl/pull/269).

File: src/context.rs
Comment (neumark) [inline src/context.rs:574]: Not sure if this as an issue: If I have a table called `my.table`, then could this incorrectly be parsed as `my` being the collection / schema? Disregard if "." is not a valid char even within quotes for a table name.

File: src/context.rs
Comment (neumark) [inline src/context.rs:959]: Apologies for bringing up spelling: "liter" -> "lighter"?

File: src/wasm_udf/data_types.rs
Comment (neumark) [inline src/wasm_udf/data_types.rs:77]: I realize this is just `clippy` rewriting the existing code, but can we make `WasmMessagePack` the default?

**Review by neumark — APPROVED** [review]
Very impressive achievement, the upgrade impact is enormous!
Congrats!!!

File: src/context.rs
Comment (gruuya) [reply src/context.rs:573]: Understood, though note that the `name` above is actually not an `ObjectName`, but an `OwnedTableReference` instead, and it's [`to_string()`](https://github.com/apache/arrow-datafusion/blob/fad77a4b278a441e343fb6caa76f6e3bcd27e248/datafusion/common/src/table_reference.rs#L107-L121) shouldn't require unquoting right?

Regardless, I'll wait for your PR to merge before merging this one.

File: src/context.rs
Comment (gruuya) [reply src/context.rs:574]: Good point. I just tried it out and indeed we there is this problem. Namely if I do
```sql
CREATE EXTERNAL TABLE \"my.table\"
...
```
The parsed reference has `default` for the catalog, but `"my` as the schema and `table"` as the table name.

File: src/context.rs
Comment (gruuya) [reply src/context.rs:959]: No worries, fixing 👍🏻

File: src/object_store/cache.rs
Comment (mildbyte) [inline src/object_store/cache.rs:422]: Wait, what happened here? Did something change in the `object_store`?

**Review by mildbyte — APPROVED** [review]

File: src/object_store/cache.rs
Comment (gruuya) [reply src/object_store/cache.rs:422]: Good question, I'm not quite sure. My initial impression is that something changed in the (`ArrowWriter` of the) `parquet` crate, but didn't dig into it further.

Comment (mildbyte) [reply src/object_store/cache.rs:422]: Ah, makes sense. I forgot this was using a Parquet file for tests.

**Pull request merged**
